src/proxy-animation.js

```diff
@@ -9,119 +9,350 @@ const nativeElementAnimate = window.Element.prototype.animate;
 const nativeAnimation = window.Animation;
 
 /**
- * Decides whether to use native Element.prototype.animate function in regular fashion or pass it to our polyfill
- *  so its current time is driven by scroll event
- * @param keyframes {Object} array of keyframe objects
- * @param options {Object} WAAPI options object
- * @returns {Function}
+ * Creates a ready promise with a pending task, to be executed at the time
+ * the promise is resolved. The promise will auto-resolve in the next animation
+ * frame, but may be resolved or rejected earlier as a result of API calls that
+ * change the state of the animation. The state of the promise
+ * (pending|resolved|rejected) may be queried to determine when a replacement
+ * promise is needed.
+ * @param details {Ojbect}
+ * @param task {function}
  */
-
 function createReadyPromise(details) {
-  let sequence = ++details.sequence;
-  let promise = details.readyPromise = new Promise((resolve, reject) => {
-    // TODO: We should actually not apply the animation until this is
-    // resolved.
-    requestAnimationFrame(() => {
-      // If this promise was replaced, this animation was aborted.
-      if (details.readyPromise == promise)
-        details.readyPromise = null;
-      if (details.aborted.has(sequence)) {
-        details.aborted.delete(sequence);
-        // Reject with a non-visible AbortError.
-        reject(new DOMException("Animation aborted", "AbortError"));
-        return;
-      }
+  let nativeResolve = undefined;
+  let nativeReject = undefined;
+  let pendingTask = undefined;
+  let pendingTaskName = undefined;
+  let state = 'pending';
+  const p = new Promise((resolve, reject) => {
+    nativeResolve = resolve;
+    nativeReject = reject;
+  });
 
-      // See pending play and pause tasks in:
-      // https://www.w3.org/TR/web-animations-1/#playing-an-animation-section
-      // https://www.w3.org/TR/web-animations-1/#pausing-an-animation-section
+  p.resolve = () => {
```

Comment:
We shouldn't modify the promise directly. We can create a PromiseWrapper class which wraps the native promise but also allows us to return the native promise from the animation methods (and still have exposed functions to resolve or reject the promise). I think this also removes a fair bit of duplication since there is very similar promise wrapping in createFinishPromise.

Reply:
done